designdocs/permissions.md

```diff
@@ -0,0 +1,87 @@
+* there may be new attacks devised or new flaws discovered after we have shipped, but user agents will not be able to address these because they are constrained by the spec.
+
+There are most likely more concerns than just what I have listed above.
+
+### Solution
+To address these I propose the following: *do both*.  Let the developers have a single code path where they /request permissions at application start/ *and* when /the permission is actually needed/. The first request is considered the upper bounds of possible permissions. The user agent can decide whether to actually show user request dialogs at app time or when the permission is actually used, or some other scheme that the user agent deems to be better (such as warding off new attacks in the future).
+
+### Details
+From the developer’s perspective it would work like this.
+
+The developer specifies the total permissions they are likely to need during the entire run of the application when requesting the XR Session, using an array with pre-defined permissions constants.  Something like this:
+
+``` javascript
+const perms = [
+   MICROPHONE,
+	 IMMERSIVE_MODE,
+   HEADSET_POSE,
+   //CAMERA_DATA,  developer decides not to use camera
+]
+navigator.xr.requestSession(perms).then(session => {
+	//start up the rest of the app
+})
+```
+
+At this point the user agent can either prompt the user for these permissions, or merely use them as an upper bound of permissions that could potentially be requested later.
+
+
+Later, when the application needs to actually use a permission, say the microphone, the developer will request this access through a promise based API. Ex:
+
+```javascript
+navigator.mediaDevices.getUserMedia(opts).then(stream => {
+	console.log("i have the microphone stream")
+})
+```
+
+At this point the user agent can either prompt the user for the microphone permission, or immediately grant it if the user already approved it previously.  
+
+Note that if the application did not include MICROPHONE as  the initial list of permissions to `requestSession` then the user agent *must* reject any later requests to `getUserMedia()`.  
+
+*The initial list of permissions is the maximum set of permissions the application may request throughout it’s lifetime.*  This enables, say, a vr chat client to let the user enter a move and talk only mode and later request camera access because the user wants to share something, without having to bug the user initially; if the user agent supports immersive permission requests.  On user agents which do not support immersive permission requests, for whatever reason (form factor, user prefs, domain name, phase of moon, etc.), then the request for camera can be done at application start.
+
+
+### Advantages
+This system has two advantages over other approaches.
+* the developer has a single code path that will always be followed.  They will not need to write conditional code to work around different user agent authentication systems.
```

Comment:
Developers will likely want a method for determining if permission for a feature has already been granted prior to calling a potentially-permission-producing method. This is to allow for the common pattern of displaying a quick in-app explanation regarding why the user is about to see a dialog just prior to displaying the platform-level prompt. If the permission has already been granted at session start, however, they would end up showing a message saying "Hey! You're about to see something that looks like this, and here's why you should click 'Yes'" followed by nothing, which appear broken to the user.

Reply:
Ah, i meant to put that other comment here... moving it....

If we do take the approach of integrating with https://www.w3.org/TR/permissions/#enumdef-permissionname then we can also take advantage of the https://www.w3.org/TR/permissions/#dom-permissions-query API?  Which to be honest, is interesting to me that this is queryable since for some reason I thought there were concerns about websites being able to differentiate between a user rejecting a permission and some other random reason a feature might not be available?  🤔

Reply:
yes, I thought this was a fingerprinting issue? If not then yes, let's have a query.

Reply:
I can't think of a decent way to use the query API as a fingerprint, but I'm happy to be corrected if someone else can. I'm tentatively in favor of allowing the query (because it will enable more progressive permissions--yay!)